File: Algorithm/2023.10/BinaryHyeok/다단계 칫솔 판매.java
import java.util.*;

class Solution {
    
    class Node{
        String name, referral;
        int profit;
        
        public Node(String enroll, String referral, int profit){
            this.name = name;
            this.referral = referral;
            this.profit = profit;
        }

    }
    public int[] solution(String[] enroll, String[] referral, String[] seller, int[] amount) {
        Map<String, Node> groupMap = new HashMap<>();
        
        // map 정보 입력
        for(int i = 0; i < enroll.length; i++){
            String name = enroll[i];
            String ref = referral[i];
            groupMap.put(name, new Node(name, ref, 0));
        }
        
        for(int i = 0; i < seller.length; i++){
            String name = seller[i];
            int price = amount[i] * 100;
            
            // 추천인에게 10%의 이익을 준다.
            while(true){
                int parentProfit = price / 10;
                int myProfit = price - parentProfit;
                
                // 자기 자신의 이익을 추가
                Node node = groupMap.get(name);
                node.profit += myProfit;
                groupMap.put(name, node);
                
                // 추천인 정보로 갱신
                name = node.referral;
                price = parentProfit;
                
                // 추천인이 없거나 줄 이익이 없으면 종료
                if("-".equals(name) || price < 1){
                    break;
                }
            }
        }
        
        int[] answer = new int[enroll.length];
        for(int i = 0; i < answer.length; i++){
            answer[i] = groupMap.get(enroll[i]).profit;
        }
        
        return answer;
    }
}

File: Algorithm/2023.10/Wolfy916/코딩테스트 기출문제/다단계 칫솔 판매.py
'''
다단계 칫솔 판매 - 프로그래머스
분류 : 연결리스트, 트리
'''

def solution(enroll, referral, seller, amount):
    # [1] 연결리스트 생성
    tree = dict()
    tree["-"] = [0, 'e']
    for i in range(len(enroll)):
        # tree[이름] = [총 이익금, 추천인 이름]
        tree[enroll[i]] = [0, referral[i]]
    
    # [2] 이익금 계산
    for i in range(len(seller)):
        user = seller[i]
        benefit = amount[i] * 100       # 판매대금
        fee = benefit // 10             # 10% 수수료
        tree[user][0] += benefit - fee  # 본인 수익금
        while fee:
            benefit = fee
            fee = benefit // 10
            par = tree[user][1]
            if par == 'e': break
            tree[par][0] += benefit - fee
            user = par
    
    # [3] 이익금 출력
    answer = []
    for user in enroll:
        answer.append(tree[user][0])
    return answer